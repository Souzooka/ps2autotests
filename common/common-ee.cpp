#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
