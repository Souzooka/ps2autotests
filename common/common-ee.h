#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
