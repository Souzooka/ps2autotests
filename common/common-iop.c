#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
