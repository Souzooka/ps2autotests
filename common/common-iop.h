#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
