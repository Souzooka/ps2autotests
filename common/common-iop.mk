# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
