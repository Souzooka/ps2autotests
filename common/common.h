#ifndef PS2AUTOTESTS_COMMON_H
#define PS2AUTOTESTS_COMMON_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#endif
