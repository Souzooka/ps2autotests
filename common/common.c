#include "common.h"
#include <stdarg.h>

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
