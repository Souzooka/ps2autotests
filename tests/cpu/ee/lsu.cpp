#include <common.h>
#include <string.h>

static u32 __attribute__((aligned(16))) C_GARBAGE1[4] = {0x1337, 0x1338, 0x1339, 0x133A};
static u32 __attribute__((aligned(16))) C_PATTERN[3][4] = {
	{0x45678123, 0x9ABCDEF0, 0xDEADBEEF, 0xC0DE1337},
	{0x23456789, 0xABCDEF01, 0xBEEFDEAD, 0xC0DEC0DE},
	{0x8899AABB, 0xCCDDEEFF, 0x00112233, 0x44556677},
};

#define LOAD_OP_FUNC(OP) \
template <int offset> \
inline void LOAD_##OP(register u128 &res, void *address) { \
	asm volatile ( \
		#OP " %0, %2(%1)\n" \
		"sync\n" \
		: "+&r"(res) : "r"(address), "i"(offset) \
	); \
} \
template <int offset> \
inline void LOADZ_##OP(register u128 &res, void *address) { \
	asm volatile ( \
		#OP " $0, %2(%1)\n" \
		"sync\n" \
		"ori %0, $0, 0\n" \
		: "+&r"(res) : "r"(address), "i"(offset) \
	); \
}

#define STORE_OP_FUNC(OP) \
template <int offset> \
inline void STORE_##OP(const register u128 &val, void *address) { \
	asm volatile ( \
		#OP " %0, %2(%1)\n" \
		"sync\n" \
		: : "r"(val), "r"(address), "i"(offset) \
	); \
}

template <u32 i>
static inline void SET_U32(register u128 &rd) {
	asm volatile (
		"lui %0, %1\n"
		"ori %0, %0, %2\n"
		: "+r"(rd) : "K"((i >> 16) & 0xFFFF), "K"(i & 0xFFFF)
	);
}

static inline void SET_M(register u128 &rd, u32 *p) {
	rd = *(vu128 *)p;
}

static inline void PRINT_R(const register u128 &rt, bool newline) {
	static u32 __attribute__((aligned(16))) result[4] = {0, 0, 0, 0};
	*(vu128 *)result = rt;

	printf("%08x %08x %08x %08x", result[3], result[2], result[1], result[0]);
	if (newline) {
		printf("\n");
	}
}

static inline void PRINT_M(const void *m, bool newline) {
	const u32 *result = (const u32 *)m;

	printf("%08x %08x %08x %08x", result[3], result[2], result[1], result[0]);
	if (newline) {
		printf("\n");
	}
}

#define LOAD_OP_DO_I(OP, t, s) \
	SET_U32<t>(rt); \
	LOAD_##OP<s>(rt, C_PATTERN[1]); \
	printf("  %s %+d: ", #OP, (u32)s); PRINT_R(rt, true);

#define LOAD_OP_DO_M(OP, t, s) \
	SET_M(rt, t); \
	LOAD_##OP<s>(rt, C_PATTERN[1]); \
	printf("  %s +%d: ", #OP, (u32)s); PRINT_R(rt, true);

#define TEST_LOAD_FUNC(OP) \
LOAD_OP_FUNC(OP); \
static void test_##OP() { \
	register u128 rt; \
	SET_M(rt, C_GARBAGE1); \
	 \
	printf("%s:\n", #OP); \
	 \
	LOAD_OP_DO_I(OP, 0xABCD4321, 0); \
	LOAD_OP_DO_I(OP, 0xABCD4321, 16); \
	LOAD_OP_DO_I(OP, 0xABCD4321, -16); \
	 \
	LOAD_OP_DO_M(OP, C_GARBAGE1, 0); \
	LOAD_OP_DO_M(OP, C_GARBAGE1, 16); \
	LOAD_OP_DO_M(OP, C_GARBAGE1, -16); \
	\
	SET_M(rt, C_GARBAGE1); \
	LOADZ_##OP<0>(rt, C_PATTERN[1]); \
	printf("  %s -> $0: ", #OP); PRINT_R(rt, true); \
	printf("\n"); \
}

#define LOADLR_OP_DO_I(LOP, ROP, t, l, r) \
	SET_U32<t>(rt); \
	LOAD_##LOP<l>(rt, C_PATTERN[1]); \
	LOAD_##ROP<r>(rt, C_PATTERN[1]); \
	printf("  %s %+d/%s %+d: ", #LOP, (u32)l, #ROP, (u32)r); PRINT_R(rt, true);

#define LOADLR_OP_DO_M(LOP, ROP, t, l, r) \
	SET_M(rt, t); \
	LOAD_##LOP<l>(rt, C_PATTERN[1]); \
	LOAD_##ROP<r>(rt, C_PATTERN[1]); \
	printf("  %s %+d/%s %+d: ", #LOP, (u32)l, #ROP, (u32)r); PRINT_R(rt, true);

#define TEST_LOADLR_FUNC(LOP, ROP, SIZE) \
LOAD_OP_FUNC(LOP); \
LOAD_OP_FUNC(ROP); \
static void test_##LOP##_##ROP() { \
	register u128 rt; \
	SET_M(rt, C_GARBAGE1); \
	 \
	printf("%s / %s:\n", #LOP, #ROP); \
	 \
	LOAD_OP_DO_I(LOP, 0xABCD4321, 0); \
	LOAD_OP_DO_I(ROP, 0xABCD4321, 0); \
	LOAD_OP_DO_I(LOP, 0xABCD4321, 1); \
	LOAD_OP_DO_I(ROP, 0xABCD4321, 1); \
	LOADLR_OP_DO_I(LOP, ROP, 0xABCD4321, 0, SIZE - 1); \
	 \
	LOAD_OP_DO_M(LOP, C_GARBAGE1, 0); \
	LOAD_OP_DO_M(ROP, C_GARBAGE1, 0); \
	LOAD_OP_DO_M(LOP, C_GARBAGE1, 1); \
	LOAD_OP_DO_M(ROP, C_GARBAGE1, 1); \
	LOADLR_OP_DO_M(LOP, ROP, C_GARBAGE1, 0, SIZE - 1); \
	\
	SET_M(rt, C_GARBAGE1); \
	LOADZ_##LOP<0>(rt, C_PATTERN[1]); \
	printf("  %s -> $0: ", #LOP); PRINT_R(rt, true); \
	SET_M(rt, C_GARBAGE1); \
	LOADZ_##ROP<0>(rt, C_PATTERN[1]); \
	printf("  %s -> $0: ", #ROP); PRINT_R(rt, true); \
	printf("\n"); \
}

#define STORE_OP_DO_I(OP, t, s) \
	memcpy(buffer, C_PATTERN, sizeof(buffer)); \
	SET_U32<t>(rt); \
	STORE_##OP<s>(rt, &buffer[1]); \
	printf("  %s %+d: ", #OP, (u32)s); PRINT_M(&buffer[1 + (s) / 16], true);

#define STORE_OP_DO_M(OP, t, s) \
	memcpy(buffer, C_PATTERN, sizeof(buffer)); \
	SET_M(rt, t); \
	STORE_##OP<s>(rt, &buffer[1]); \
	printf("  %s +%d: ", #OP, (u32)s); PRINT_M(&buffer[1 + (s) / 16], true);

#define TEST_STORE_FUNC(OP) \
STORE_OP_FUNC(OP); \
static void test_##OP() { \
	u128 buffer[3]; \
	register u128 rt; \
	SET_M(rt, C_GARBAGE1); \
	 \
	printf("%s:\n", #OP); \
	 \
	STORE_OP_DO_I(OP, 0xABCD4321, 0); \
	STORE_OP_DO_I(OP, 0xABCD4321, 16); \
	STORE_OP_DO_I(OP, 0xABCD4321, -16); \
	 \
	STORE_OP_DO_M(OP, C_GARBAGE1, 0); \
	STORE_OP_DO_M(OP, C_GARBAGE1, 16); \
	STORE_OP_DO_M(OP, C_GARBAGE1, -16); \
	\
	printf("\n"); \
}

#define STORELR_OP_DO_I(LOP, ROP, t, l, r) \
	memcpy(buffer, C_PATTERN, sizeof(buffer)); \
	SET_U32<t>(rt); \
	STORE_##LOP<l>(rt, &buffer[1]); \
	STORE_##ROP<r>(rt, &buffer[1]); \
	printf("  %s %+d/%s %+d: ", #LOP, (u32)l, #ROP, (u32)r); PRINT_M(&buffer[1 + (l) / 16], true);

#define STORELR_OP_DO_M(LOP, ROP, t, l, r) \
	memcpy(buffer, C_PATTERN, sizeof(buffer)); \
	SET_M(rt, t); \
	STORE_##LOP<l>(rt, &buffer[1]); \
	STORE_##ROP<r>(rt, &buffer[1]); \
	printf("  %s %+d/%s %+d: ", #LOP, (u32)l, #ROP, (u32)r); PRINT_M(&buffer[1 + (l) / 16], true);

#define TEST_STORELR_FUNC(LOP, ROP, SIZE) \
STORE_OP_FUNC(LOP); \
STORE_OP_FUNC(ROP); \
static void test_##LOP##_##ROP() { \
	u128 buffer[3]; \
	register u128 rt; \
	SET_M(rt, C_GARBAGE1); \
	 \
	printf("%s / %s:\n", #LOP, #ROP); \
	 \
	STORE_OP_DO_I(LOP, 0xABCD4321, 0); \
	STORE_OP_DO_I(ROP, 0xABCD4321, 0); \
	STORE_OP_DO_I(LOP, 0xABCD4321, 1); \
	STORE_OP_DO_I(ROP, 0xABCD4321, 1); \
	STORELR_OP_DO_I(LOP, ROP, 0xABCD4321, 0, SIZE - 1); \
	 \
	STORE_OP_DO_M(LOP, C_GARBAGE1, 0); \
	STORE_OP_DO_M(ROP, C_GARBAGE1, 0); \
	STORE_OP_DO_M(LOP, C_GARBAGE1, 1); \
	STORE_OP_DO_M(ROP, C_GARBAGE1, 1); \
	STORELR_OP_DO_M(LOP, ROP, C_GARBAGE1, 0, SIZE - 1); \
	\
	printf("\n"); \
}

TEST_LOAD_FUNC(lb)
TEST_LOAD_FUNC(lbu)
TEST_LOAD_FUNC(ld)
TEST_LOADLR_FUNC(ldl, ldr, 8)
TEST_LOAD_FUNC(lh)
TEST_LOAD_FUNC(lhu)
TEST_LOAD_FUNC(lw)
TEST_LOADLR_FUNC(lwl, lwr, 4)
TEST_LOAD_FUNC(lwu)
TEST_LOAD_FUNC(lq)

TEST_STORE_FUNC(sb)
TEST_STORE_FUNC(sd)
TEST_STORELR_FUNC(sdl, sdr, 8)
TEST_STORE_FUNC(sh)
TEST_STORE_FUNC(sw)
TEST_STORELR_FUNC(swl, swr, 4)
TEST_STORE_FUNC(sq)

static void test_pref() {
	// No guarantees that it will even prefetch, let's just make sure it doesn't wig out.
	register u32 res = 0;
	asm volatile (
		"pref 0, 0(%1)\n"
		"ori %0, $0, 0\n"
		: "+&r"(res) : "r"(C_PATTERN)
	);
	printf("pref: %s\n", res == 0 ? "OK" : "FAIL");
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	test_lb();
	test_lbu();
	test_ld();
	test_ldl_ldr();
	test_lh();
	test_lhu();
	test_lw();
	test_lwl_lwr();
	test_lwu();
	test_lq();

	test_sb();
	test_sd();
	test_sdl_sdr();
	test_sh();
	test_sw();
	test_swl_swr();
	test_sq();

	test_pref();

	printf("-- TEST END\n");

	return 0;
}
