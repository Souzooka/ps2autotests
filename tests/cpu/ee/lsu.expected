-- TEST BEGIN
lb:
  lb +0: 0000133a 00001339 ffffffff ffffff89
  lb +16: 0000133a 00001339 ffffffff ffffffbb
  lb -16: 0000133a 00001339 00000000 00000023
  lb +0: 0000133a 00001339 ffffffff ffffff89
  lb +16: 0000133a 00001339 ffffffff ffffffbb
  lb +-16: 0000133a 00001339 00000000 00000023
  lb -> $0: 0000133a 00001339 00000000 00000000

lbu:
  lbu +0: 0000133a 00001339 00000000 00000089
  lbu +16: 0000133a 00001339 00000000 000000bb
  lbu -16: 0000133a 00001339 00000000 00000023
  lbu +0: 0000133a 00001339 00000000 00000089
  lbu +16: 0000133a 00001339 00000000 000000bb
  lbu +-16: 0000133a 00001339 00000000 00000023
  lbu -> $0: 0000133a 00001339 00000000 00000000

ld:
  ld +0: 0000133a 00001339 abcdef01 23456789
  ld +16: 0000133a 00001339 ccddeeff 8899aabb
  ld -16: 0000133a 00001339 9abcdef0 45678123
  ld +0: 0000133a 00001339 abcdef01 23456789
  ld +16: 0000133a 00001339 ccddeeff 8899aabb
  ld +-16: 0000133a 00001339 9abcdef0 45678123
  ld -> $0: 0000133a 00001339 00000000 00000000

ldl / ldr:
  ldl +0: 0000133a 00001339 89ffffff abcd4321
  ldr +0: 0000133a 00001339 abcdef01 23456789
  ldl +1: 0000133a 00001339 6789ffff abcd4321
  ldr +1: 0000133a 00001339 ffabcdef 01234567
  ldl +0/ldr +7: 0000133a 00001339 89ffffff abcd43ab
  ldl +0: 0000133a 00001339 89001338 00001337
  ldr +0: 0000133a 00001339 abcdef01 23456789
  ldl +1: 0000133a 00001339 67891338 00001337
  ldr +1: 0000133a 00001339 00abcdef 01234567
  ldl +0/ldr +7: 0000133a 00001339 89001338 000013ab
  ldl -> $0: 0000133a 00001339 00000000 00000000
  ldr -> $0: 0000133a 00001339 00000000 00000000

lh:
  lh +0: 0000133a 00001339 00000000 00006789
  lh +16: 0000133a 00001339 ffffffff ffffaabb
  lh -16: 0000133a 00001339 ffffffff ffff8123
  lh +0: 0000133a 00001339 00000000 00006789
  lh +16: 0000133a 00001339 ffffffff ffffaabb
  lh +-16: 0000133a 00001339 ffffffff ffff8123
  lh -> $0: 0000133a 00001339 00000000 00000000

lhu:
  lhu +0: 0000133a 00001339 00000000 00006789
  lhu +16: 0000133a 00001339 00000000 0000aabb
  lhu -16: 0000133a 00001339 00000000 00008123
  lhu +0: 0000133a 00001339 00000000 00006789
  lhu +16: 0000133a 00001339 00000000 0000aabb
  lhu +-16: 0000133a 00001339 00000000 00008123
  lhu -> $0: 0000133a 00001339 00000000 00000000

lw:
  lw +0: 0000133a 00001339 00000000 23456789
  lw +16: 0000133a 00001339 ffffffff 8899aabb
  lw -16: 0000133a 00001339 00000000 45678123
  lw +0: 0000133a 00001339 00000000 23456789
  lw +16: 0000133a 00001339 ffffffff 8899aabb
  lw +-16: 0000133a 00001339 00000000 45678123
  lw -> $0: 0000133a 00001339 00000000 00000000

lwl / lwr:
  lwl +0: 0000133a 00001339 ffffffff 89cd4321
  lwr +0: 0000133a 00001339 00000000 23456789
  lwl +1: 0000133a 00001339 00000000 67894321
  lwr +1: 0000133a 00001339 ffffffff ab234567
  lwl +0/lwr +3: 0000133a 00001339 ffffffff 89cd4323
  lwl +0: 0000133a 00001339 ffffffff 89001337
  lwr +0: 0000133a 00001339 00000000 23456789
  lwl +1: 0000133a 00001339 00000000 67891337
  lwr +1: 0000133a 00001339 00001338 00234567
  lwl +0/lwr +3: 0000133a 00001339 ffffffff 89001323
  lwl -> $0: 0000133a 00001339 00000000 00000000
  lwr -> $0: 0000133a 00001339 00000000 00000000

lwu:
  lwu +0: 0000133a 00001339 00000000 23456789
  lwu +16: 0000133a 00001339 00000000 8899aabb
  lwu -16: 0000133a 00001339 00000000 45678123
  lwu +0: 0000133a 00001339 00000000 23456789
  lwu +16: 0000133a 00001339 00000000 8899aabb
  lwu +-16: 0000133a 00001339 00000000 45678123
  lwu -> $0: 0000133a 00001339 00000000 00000000

lq:
  lq +0: c0dec0de beefdead abcdef01 23456789
  lq +16: 44556677 00112233 ccddeeff 8899aabb
  lq -16: c0de1337 deadbeef 9abcdef0 45678123
  lq +0: c0dec0de beefdead abcdef01 23456789
  lq +16: 44556677 00112233 ccddeeff 8899aabb
  lq +-16: c0de1337 deadbeef 9abcdef0 45678123
  lq -> $0: 0000133a 00001339 00000000 00000000

sb:
  sb +0: c0dec0de beefdead abcdef01 23456721
  sb +16: 44556677 00112233 ccddeeff 8899aa21
  sb -16: c0de1337 deadbeef 9abcdef0 45678121
  sb +0: c0dec0de beefdead abcdef01 23456737
  sb +16: 44556677 00112233 ccddeeff 8899aa37
  sb +-16: c0de1337 deadbeef 9abcdef0 45678137

sd:
  sd +0: c0dec0de beefdead ffffffff abcd4321
  sd +16: 44556677 00112233 ffffffff abcd4321
  sd -16: c0de1337 deadbeef ffffffff abcd4321
  sd +0: c0dec0de beefdead 00001338 00001337
  sd +16: 44556677 00112233 00001338 00001337
  sd +-16: c0de1337 deadbeef 00001338 00001337

sdl / sdr:
  sdl +0: c0dec0de beefdead abcdef01 234567ff
  sdr +0: c0dec0de beefdead ffffffff abcd4321
  sdl +1: c0dec0de beefdead abcdef01 2345ffff
  sdr +1: c0dec0de beefdead ffffffab cd432189
  sdl +0/sdr +7: c0dec0de beefdead 21cdef01 234567ff
  sdl +0: c0dec0de beefdead abcdef01 23456700
  sdr +0: c0dec0de beefdead 00001338 00001337
  sdl +1: c0dec0de beefdead abcdef01 23450000
  sdr +1: c0dec0de beefdead 00133800 00133789
  sdl +0/sdr +7: c0dec0de beefdead 37cdef01 23456700

sh:
  sh +0: c0dec0de beefdead abcdef01 23454321
  sh +16: 44556677 00112233 ccddeeff 88994321
  sh -16: c0de1337 deadbeef 9abcdef0 45674321
  sh +0: c0dec0de beefdead abcdef01 23451337
  sh +16: 44556677 00112233 ccddeeff 88991337
  sh +-16: c0de1337 deadbeef 9abcdef0 45671337

sw:
  sw +0: c0dec0de beefdead abcdef01 abcd4321
  sw +16: 44556677 00112233 ccddeeff abcd4321
  sw -16: c0de1337 deadbeef 9abcdef0 abcd4321
  sw +0: c0dec0de beefdead abcdef01 00001337
  sw +16: 44556677 00112233 ccddeeff 00001337
  sw +-16: c0de1337 deadbeef 9abcdef0 00001337

swl / swr:
  swl +0: c0dec0de beefdead abcdef01 234567ab
  swr +0: c0dec0de beefdead abcdef01 abcd4321
  swl +1: c0dec0de beefdead abcdef01 2345abcd
  swr +1: c0dec0de beefdead abcdef01 cd432189
  swl +0/swr +3: c0dec0de beefdead abcdef01 214567ab
  swl +0: c0dec0de beefdead abcdef01 23456700
  swr +0: c0dec0de beefdead abcdef01 00001337
  swl +1: c0dec0de beefdead abcdef01 23450000
  swr +1: c0dec0de beefdead abcdef01 00133789
  swl +0/swr +3: c0dec0de beefdead abcdef01 37456700

sq:
  sq +0: 0000133a 00001339 ffffffff abcd4321
  sq +16: 0000133a 00001339 ffffffff abcd4321
  sq -16: 0000133a 00001339 ffffffff abcd4321
  sq +0: 0000133a 00001339 00001338 00001337
  sq +16: 0000133a 00001339 00001338 00001337
  sq +-16: 0000133a 00001339 00001338 00001337

pref: OK
-- TEST END
